### DISCUSSION TITLE
Questions on UDP-In/Output with multiple Control Surfaces

### DISCUSSION THREAD

**Discussion by Eriklyy** [ImperialCollegeLondon/sharpy#159] [Q&A]
Hi,

I'm playing at the moment a little bit around with the UDP-In/Output and struggle with some small issues. 

1. My first issue is propably not an issue with the UDP, than more with the settings before. I'm trying to have two control deflection inputs, coming from an external python script via UDP. The values are read into sharpy and get the right assignment to the variable names, but the script struggles with updating the second control surface. With trying (assigning both receiving values to the same variable) a figured out that this issue is based on the fact, that there is only one input field in control_surface_deflection in the timestep information (see at the bottom of the picture), even if there are two control surfaces declared in the aero settings (see picture at top). Could you please let me know which settings in have to change to create the second control_surface_deflection input in the timestep_info?
2. While trying to export the variables 'strain' and 'loads' the script has also some struggles. Since both of them are variables of safed in 'postproc_cell', I assumed that the problem is based on the fact, that these variables just don't exist at the moment when the UDP-Output requires them. Is this assumption correct (or am I just failing the export)? If yes do you have an idea for a smart workaround? 

I hope I explained my problems understandable, if not feel free to ask for more information.

![Screenshot_UDP_Problem](https://user-images.githubusercontent.com/84858758/119681551-741b5780-be42-11eb-8206-e22a2edffcfd.JPG)

**Comment by ngoiz** [comment]
Erik, you are correct in all your queries so I'm going to try and point out locations in the code for you to try and modify and add these capabilities to SHARPy (this is the beauty of open source!)

First, I did a quick modification very recently (actually in line with one of the threads in the forum #149) to add support for element variables like the cartesian rotation vector. The branch is yet to be merged but it is [fix_udp_crv](https://github.com/ImperialCollegeLondon/sharpy/tree/fix_udp_crv), hopefully to be merged soon. I would recommend you fork the repo, work on it and then once you add these capabilities do a pull request so we can add it in to the main repo. The best would be to keep the PRs as simple as possible and focused on a specific new feature (i.e. multiple control surfaces and the `postproc_cell` variables, for instance).

That said, places to look for to fix your issues:
1. For the control surface, the class `Aerogrid` is what defines the UVLM lattice and the control surfaces do their magic there. I think that is already properly set up for multiple controlled (type=2) control surfaces: https://github.com/ImperialCollegeLondon/sharpy/blob/52960e4fcf2d1bd6babcacf2240bdb5f4e36336e/sharpy/aero/models/aerogrid.py#L262

I think the issue is likely to live here, in the UDP variable setter, where it is replacing the array with a single value...
https://github.com/ImperialCollegeLondon/sharpy/blob/52960e4fcf2d1bd6babcacf2240bdb5f4e36336e/sharpy/io/inout_variables.py#L147

2. The output should be relatively easy to add. Here is where element variables are 'found':
https://github.com/ImperialCollegeLondon/sharpy/blob/52960e4fcf2d1bd6babcacf2240bdb5f4e36336e/sharpy/io/inout_variables.py#L75
I would add a `try, except` block to check whether the desired variable is in the `StructTimeStepInfo` class and if not, check in the `postproc_cell[variable]` dictionary and if then it cannot find it then raise the error.

Word of warning...the UDP method is based on multithreading within the `DynamicCoupled` solver, one thread deals with the simulation and another with the I/O. It most likely will be my programming skills falling short, but if there is an error on one of the threads it will not kill the program, it will just freeze. Therefore, if you are debugging/developing use the `logger` tool present in the UDP modules to write out the output, don't rely on errors or `print()` messages because they may never appear. If you find a way to improve this then please let me know :) !

Hope this helps

PS: whilst you are at it, if you see ANYTHING that needs changing/does not make sense/is poorly explained/could be done better, please feel free to modify it. Don't feel as if changes need to be minimal one liners, the better we make this, the merrier.

**Reply by ngoiz** [reply]
Just to note that the behaviour has changed slightly from the initial approach to sending element variables via UDP. Now, element variables are treated just like nodal variables, making everything simpler.

Full details in #160 which is now part of `develop`

**Reply by Eriklyy** [reply]
Thank you so much for your answer.

1.)You were right, the first problem was a problem of the UDP variable setter. I changed it in a way that it should automatically adjust the lenght to the requested number of control surfaces. However this has now the peculiarity, that you need to order the control surfaces inputs in the right order. F.e. if you have 3 control surfaces inputs, you need the inputs for the first and the second before the input of the third, even if you dont want to control the first two. Theoretically there should be an logger.error which should appear and give an hint to do it in the right order, if it wasnt done at first.

2.) The second issue is still an issue. I tried to do it also with try and except and it still hasnt found the variable. So I decieded to print the dictionarys of the 'timestep_info' and 'postproc_cell'.  The dictionary of the timestepinfo (marked in blue at the picture) has the attribute 'postproc_cell' which is good, but in the dictionary of postproc_cell (marked in red) arent the variables loads/strain.
Kinda unsure how to solve that, or how important it will be at the end to have access to this variables.

![Screenshot_dir_timestep_info](https://user-images.githubusercontent.com/84858758/120355621-b8579d80-c303-11eb-914d-82a6e93a235e.JPG)

PS: Both changes aren't done in the new fix_udp_crv branch, but especially the first point should be still be very easy to add. Should I add it in this one and do than a pull request?

**Reply by ngoiz** [reply]
Hi Erik, 

Good to hear that the control surfaces have been fixed, thanks for that. Maybe we can look for an option in the future where the controlled control surfaces (those with `type==2`) are only needed or work on them based on their `control_surface_id` but it's great that we now have this extra functionality.

As for point 2, I think it's an easy fix :) The `postproc_cell`/`postproc_node` dictionaries only get populated by postprocessors (most of the time). So, in order for you to have loads and strains in the `postproc_cell` dictionary you need to add the `BeamLoads` postprocessor to the ones that are run in `DynamicCoupled`, in the same way `BeamPlot` is called for instance. So your `DynamicCoupled['postprocessors']= ['BeamLoads', 'BeamPlot']` for example. In this case, in the paraview file you would also see strains and loads since the BeamLoads postproc has run before BeamPlot.

A few notes from the screenshot you attach. Since `StructTimestepInfo` is a class with the variables as attributes we need `getattr(tstep, var_name)` to fetch it. On the other hand `postproc_cell` is a dictionary so in order to find your variable you would need to use simply `tstep['postproc_cell'].get(var_name, <default_if_KeyError>)`. Also, try to avoid nonspecific `try, except` blocks. In your screenshot, in line 65 what your are looking for is an `AttributeError` so add that better in the exception clause (in case something that is not an `AttributeError` breaks it).

The `fix_udp_branch` has ceased to exist :) It was merged into develop in #160. What you would need to do in your fork is get the `develop` branch which has the latest in the UDP and work off that. Once you have your feature ready then you can do the PR.

hope this helps
